src/java.base/share/classes/sun/security/pkcs/NamedPKCS8Key.java

```diff
@@ -0,0 +1,143 @@
+ * by Oracle in the LICENSE file that accompanied this code.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+package sun.security.pkcs;
+
+import sun.security.util.DerInputStream;
+import sun.security.util.DerValue;
+import sun.security.x509.AlgorithmId;
+
+import javax.security.auth.DestroyFailedException;
+import java.io.IOException;
+import java.io.InvalidObjectException;
+import java.io.ObjectInputStream;
+import java.io.Serial;
+import java.security.InvalidKeyException;
+import java.security.NoSuchAlgorithmException;
+import java.security.ProviderException;
+import java.security.spec.NamedParameterSpec;
+import java.util.Arrays;
+
+/// Represents a private key from an algorithm family that is specialized
+/// with a named parameter set.
+///
+/// This key is generated by either a [sun.security.provider.NamedKeyPairGenerator]
+/// or [sun.security.provider.NamedKeyFactory]. Its [#getAlgorithm] method
+/// returns the algorithm family name, while its [#getParams()] method returns
+/// the parameter set name as a [NamedParameterSpec] object. The algorithm
+/// identifier in the PKCS #8 encoding of the key is always a single OID derived
+/// from the parameter set name.
+///
+/// @see sun.security.provider.NamedKeyPairGenerator
+public final class NamedPKCS8Key extends PKCS8Key {
+    @Serial
+    private static final long serialVersionUID = 1L;
+
+    private final String fname;
+    private final transient NamedParameterSpec paramSpec;
+    private final byte[] rawBytes;
+
+    private transient boolean destroyed = false;
+
+    /// Ctor from family name, parameter set name, raw key bytes.
+    /// Key bytes won't be cloned, caller must relinquish ownership
+    public NamedPKCS8Key(String fname, String pname, byte[] rawBytes) {
+        this.fname = fname;
+        this.paramSpec = new NamedParameterSpec(pname);
+        try {
+            this.algid = AlgorithmId.get(pname);
+        } catch (NoSuchAlgorithmException e) {
+            throw new ProviderException(e);
+        }
+        this.rawBytes = rawBytes;
+
+        DerValue val = new DerValue(DerValue.tag_OctetString, rawBytes);
+        try {
+            this.key = val.toByteArray();
+        } finally {
+            val.clear();
+        }
```

Comment:
Yes, `this.key` is the one inside `PKCS8Key`.

Since EdDSA and XDH, the private key has taken this OCTET in OCTET approach. My code is identical to what you see at https://github.com/openjdk/jdk/blob/adca97b659d725b0dd320322297dcbd1b443a047/src/java.base/share/classes/sun/security/ec/ed/EdDSAPrivateKeyImpl.java#L50.

In https://datatracker.ietf.org/doc/html/rfc8410#autoid-7 and https://www.ietf.org/archive/id/draft-ietf-lamps-kyber-certificates-04.html#name-private-key-format, you can see the definitions:

```
   OneAsymmetricKey ::= SEQUENCE {
      version Version,
      privateKeyAlgorithm PrivateKeyAlgorithmIdentifier,
      privateKey PrivateKey,
      ...
   }

   PrivateKey ::= OCTET STRING

   ... CurvePrivateKey object and wrapped by the OCTET STRING of the
   "privateKey" field.

   CurvePrivateKey ::= OCTET STRING
```

Reply:
For what it's worth, my [PEM](https://github.com/openjdk/jdk/pull/17543) changes to [PKCS8Key.java](https://github.com/openjdk/jdk/pull/17543/files#diff-d4d775f071342d20e524e55883168e018a15f32e0d607518ef3d5f0f76dcdd29) change `key` to `privKeyMaterial` because `key` doesn't refer to the Key interface, but binary data

Reply:
@ascarpino Yes I noticed it when I was trying to apply your patch onto mine to see if I can use PEM to encode and decode my new keys. Let's see who gets integrated first. :-)